### PR TITLE
PD-583: Add client-side filtering to ProjectSelect and OrgSelect

### PR DESCRIPTION
**Pull request by brookescarlett**
## ✍️ Proposed changes

- Add client-side filtering to ProjectSelect and OrgSelect

🎟 _Jira ticket:_ [PD-583](https://jira.mongodb.org/browse/PD-583)

## 🛠 Types of changes

- [ ] Tooling (updates to workspace config or internal tooling)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

### CODE REVIEW

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx [hunk 1]
```diff
@@ -115,28 +115,61 @@ interface OrganizationMongoSelectProps extends BaseMongoSelectProps {
...
+
+  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setValue(e.target.value);
+
+    if (onChangeProp) {
+      onChangeProp(e);
+      return;
+    }
+
+    const search = new RegExp(`${e.target.value}`, 'i');
```
Comment (hswolff) [inline packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: If you're using the same value (`e.target.value`) twice in a scope it's generally good practice to cache that value in a local variable.

This is a micro performance optimization, as every property access takes some (negligible) cost to do, but most importantly it guarantees the consistency of that value used throughout the function.

Such that you sanitize the value into a local variable at the beginning of the function and then delegate all usage to that one source of (consistent) truth.

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx [hunk 2]
```diff
@@ -115,28 +115,61 @@ interface OrganizationMongoSelectProps extends BaseMongoSelectProps {
   isOnPrem?: boolean;
 }
 
-const onKeyDown: React.KeyboardEventHandler = e => {
```
Comment (hswolff) [inline packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: You can  definitely update this function so you can reuse it across both selectors. Seems important to guarantee the similarity in behavior.

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx [hunk 3]
```diff
@@ -115,28 +115,61 @@ interface OrganizationMongoSelectProps extends BaseMongoSelectProps {
...
 
+  useEffect(() => {
+    setRenderedData(data);
+  }, [data]);
+
+  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setValue(e.target.value);
+
+    if (onChangeProp) {
+      onChangeProp(e);
```
Comment (hswolff) [inline packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: So basically we only filter internally if the user doesn't set `onChangeProp`, right?

I think two things are possible to improve here:

- Abstract the filter logic into a standalone function that can be shared
- Not have to create an intermediate state value of `renderedData`. I think you can do this all inline in the component body.

Happy to pair on this.

File: packages/mongo-nav/src/mongo-select/MongoSelect.spec.tsx [hunk 4]
```diff
@@ -175,6 +172,40 @@ describe('packages/mongo-select', () => {
         expect(orgSettingsIcon).toBeNull();
       });
     });
+
+    describe('when a user filters organizations', () => {
```
Comment (hswolff) [inline packages/mongo-nav/src/mongo-select/MongoSelect.spec.tsx]: Need a test for when a user sets their `onChange` handler to make sure we don't filter.

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx (hunk 1 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: 👍

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx (hunk 2 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: 👍

File: packages/mongo-nav/src/mongo-select/MongoSelect.spec.tsx (hunk 4 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/mongo-select/MongoSelect.spec.tsx]: 👍

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx [hunk 5]
```diff
@@ -115,28 +132,48 @@ interface OrganizationMongoSelectProps extends BaseMongoSelectProps {
...
   loading = false,
 }: OrganizationMongoSelectProps) {
+  const [value, setValue] = useState('');
   const [open, setOpen] = useState(false);
   const onElementClick = useOnElementClick();
 
+  let renderedData = data;
+
+  const filterData = () => {
+    const search = new RegExp(`${value}`, 'i');
```
Comment (brookescarlett) [inline packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: Known bug if you typeL "\" it breaks, should I not turn the value into a regex?

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx (hunk 3 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: 👍

File: packages/mongo-nav/src/org-nav/OrgNav.tsx [hunk 6]
```diff
@@ -106,22 +106,21 @@ const paymentStatusMap: Record<Colors, ReadonlyArray<OrgPaymentLabel>> = {
   ],
 } as const;
 
-interface OrgNav {
+type OrgNavProps = Pick<
```
Comment (hswolff) [inline packages/mongo-nav/src/org-nav/OrgNav.tsx]: Niiiicee!!

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx (hunk 5 above)
Comment (hswolff) [reply packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: Can you not just do `new RegExp(String(value), 'i')` ?

Do we have tests for these cases?

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx [hunk 7]
```diff
@@ -92,11 +92,28 @@ const formattedPlanTypes: Record<PlanType, string> = {
   [PlanType.OnPrem]: 'Ops Manager',
 } as const;
 
+const onKeyDown = (e: React.KeyboardEvent, callback: Function) => {
```
Comment (hswolff) [inline packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: The callback will always be the useState setter function, so we can just pass that in directly.

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx (hunk 5 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: Yes the simplification works but still fails on \

File: packages/mongo-nav/src/org-nav/OrgNav.tsx (hunk 6 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/org-nav/OrgNav.tsx]: 👍

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx (hunk 7 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: What do you mean?

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx [hunk 8]
```diff
@@ -92,11 +92,28 @@ const formattedPlanTypes: Record<PlanType, string> = {
   [PlanType.OnPrem]: 'Ops Manager',
 } as const;
 
+const onKeyDown = (e: React.KeyboardEvent, setValue: Function) => {
```
Comment (hswolff) [inline packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: Heh, misunderstood.

This should literally be the setValue function from useState.

Cuz then in the body you can just do `setValue(currentValue => `${currentValue} `)`. That usage is consistent everywhere.

Comment (brookescarlett) [reply packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: 🤦‍♀ got it

**Pull request merged**